### PR TITLE
Add optional Rustls support

### PR DESCRIPTION
**Pull request by stefankreutz**
Hello. I would like to discuss adding optional [Rustls](https://github.com/rustls/rustls) support to Lychee. I've prepared a pull request to show that's it's mostly a question of passing feature flags down to existing dependencies. What do you think?

---

This commit adds a non-default feature flag to use Rustls instead of OpenSSL.

My personal motivation is to use Lychee on OpenBSD -current, where the `openssl` crate frequently fails to link against the unreleased system LibreSSL. Using the `vendored-openssl` feature helps with compilation, but segfaults at runtime.

The commit adds three feature flags to the library, binary, benchmark, and all examples:

- The `native-tls` feature flag toggles the `openssl` crate.
- The `rustls-tls` feature flag toggles the `rustls` crate.
- The `email-check` feature flag toggles the `check-if-email-exists` crate, which is the only existing functionality currently incompatible with Rustls.

By default, `native-tls` and `email-check` are enabled. Thus, Lychee (bin and lib) can be used as before unless default features are disabled.

To use the Rustls feature, pass `--no-default-features --features rustls` to cargo check/build/test/..., e.g.,

    $ cargo clippy --workspace --all-targets --no-default-features --features rustls-tls -- --deny warnings

Checking email addresses requires both, `native-tls` and `email-check`, to be enabled. Otherwise, email addresses are excluded.

The `email-check` feature flag is technically not necessary. I preferred it over `not(rustls-tls)` because it's clearer and it addresses the AGPL license issue #594. As far as I understand, a Lychee binary compiled without the `email-check` feature could be distributed with file-based copyleft for the MPL-licensed dependencies only. But that's out of scope here.

The benchmark shows a performance regression varying between 2% and 4.4% when using Rustls instead of OpenSSL on my machine.

PS: The `ring` crate needs to be patched on OpenBSD 7.3 and later until the new xonly patches have been upstreamed, see the `rust-ring` port.

### CODE REVIEW

**Comment by mre** [comment]
Couldn't take a look at the code yet, but I like the idea and I'm all for supporting more platforms. Thanks for the efforts so far.

File: lychee-bin/Cargo.toml [hunk 1]
```diff
@@ -24,10 +24,10 @@ const_format = "0.2.31"
 headers = "0.3.8"
 http = "0.2.9"
 indicatif = "0.17.5"
-openssl-sys = "0.9.88"
+openssl-sys = { version = "0.9.88", optional = true }
 pad = "0.1.6"
 regex = "1.8.4"
-reqwest = { version = "0.11.18", features = ["gzip"] }
+reqwest = { version = "0.11.18", default-features = false, features = ["gzip", "json"] }
```
Comment (mre) [inline lychee-bin/Cargo.toml:30]: Interesting. Was the json feature missing before? Was it somehow implicitly activated?

File: lychee-lib/Cargo.toml [hunk 2]
```diff
@@ -17,17 +17,17 @@ repository = "https://github.com/lycheeverse/lychee"
...
 http = "0.2.9"
 linkify = "0.9.0"
-openssl-sys = "0.9.88"
+openssl-sys = { version = "0.9.88", optional = true }
 pulldown-cmark = "0.9.3"
 regex = "1.8.4"
 # Use trust-dns to avoid lookup failures on high concurrency
 # https://github.com/seanmonstar/reqwest/issues/296
-reqwest = { version = "0.11.18", features = ["gzip", "trust-dns"] }
+reqwest = { version = "0.11.18", default-features = false, features = ["gzip", "trust-dns"] }
```
Comment (mre) [inline lychee-lib/Cargo.toml:30]: This could potentially save a bit of compilation time.

**Review by mre — APPROVED** [review]
Changes look good. I wonder if we should mention something in the docs (maybe a line or two in the `README.md` somewhere)?

File: lychee-bin/Cargo.toml (hunk 1 above)
Comment (stefankreutz) [reply lychee-bin/Cargo.toml:30]: Good question. I assumed "json" was a default feature of reqwest, but it is not. In fact, I cannot see how it worked before. There is no other reqwest feature that enables the "json" feature, and it is necessary for the `lychee::archive::wayback::get_wayback_link` function.

**Review by mre — APPROVED** [review]
Awesome, thanks 🙏

**Pull request merged**